src/adapter/src/coord.rs

```diff
@@ -2858,6 +2871,32 @@ impl Coordinator {
                     () = self.controller.ready() => {
                         messages.push(Message::ControllerReady);
                     }
+                    // `tick` documented to be cancellation safe.
+                    // Receive a regular maintenance message.
+                    _ = maintenance_ticker.tick() => {
+                        messages.push(Message::Maintenance);
+                        maintenance_ticker.reset();
+                    }
+                    // `recv_many()` on `UnboundedReceiver` is cancellation safe:
+                    // https://docs.rs/tokio/1.38.0/tokio/sync/mpsc/struct.UnboundedReceiver.html#cancel-safety-1
+                    // Receive a batch of compute controller responses.
+                    count = self.controller_rxs.compute_controller_rx.recv_many(&mut compute_messages, MESSAGE_BATCH) => {
+                        if count == 0 {
+                            break;
+                        } else {
+                            messages.extend(compute_messages.drain(..).map(Message::Compute));
+                        }
+                    }
+                    // `recv_many()` on `UnboundedReceiver` is cancellation safe:
+                    // https://docs.rs/tokio/1.38.0/tokio/sync/mpsc/struct.UnboundedReceiver.html#cancel-safety-1
+                    // Receive a batch of controller responses.
+                    count = self.controller_rxs.immediate_rx.recv_many(&mut controller_messages, MESSAGE_BATCH) => {
+                        if count == 0 {
+                            break;
+                        } else {
+                            messages.extend(controller_messages.drain(..).map(Message::Controller));
+                        }
+                    }
```

Comment:
Didn't review this too in-depth yet but my initial thought is including all three of these new branches in the `select!` feels like a bit of an abstraction leak? Do you think it's possible to bundle these two channels and the ticker into `ControllerReceivers` and we can await with via a single `async fn process(...) -> Vec<...>` method?

Reply:
The simple version of Parker's suggestion would be to simply let the existing `Controller::process` return a `Vec<ControllerResponse>` instead. I think I'd prefer that since presumably it ends up less complicated than having a `ControllerReceivers::process` (or more likely `ControllerReceivers::recv`, to indicate that the method needs to be cancel safe).

I like getting rid of the `ready`/`process` pattern in favor of a channel, but I also see that adapter doesn't want to have to worry about have compute controller responses are handled in the controller. I think ideally, the controller would be a set of tasks and a client facade, and controller responses would simply arrive on a channel with no expectation that they are fed back into the controller (i.e. only the `immediate_rx` would exist here). But getting there still needs a lot of refactoring. So far, only the compute controller has separate tasks, while both the storage controller and the top-level `Controller` want to perform their own processing in the coordinator task and therefore need to sometimes have "process"/"maintain" methods invoked by the coordinator.

I think it will be instructive to have performance measurements. If it turns out that the approach implemented here performs better than both the current implementation and the `fn process(...) -> Vec<...>` approach, it will make it more worthwhile to think about ways to avoid the leaky abstraction (or simply accept it).